### PR TITLE
Remove boost::program_options dependency

### PR DESCRIPTION
**Pull request by ken2812221**
Concept from #12744, but without parsing negated options.

### CODE REVIEW

File: src/util.h [hunk 1]
```diff
@@ -282,6 +282,17 @@ class ArgsManager
      * Check whether we know of this arg
      */
     bool IsArgKnown(const std::string& key, std::string& error);
+
+private:
+    /**
+     * Trim string by the pattern
+     */
+    std::string trimString(std::string str, const std::string& pattern);
```
Comment (qmma70) [inline src/util.h]: Is there a reason for this method to start with a lower case letter while all other start with upper case letters?

File: src/util.cpp [hunk 2]
```diff
@@ -818,10 +817,11 @@ bool ArgsManager::ReadConfigStream(std::istream& stream, std::string& error, boo
     std::set<std::string> setOptions;
     setOptions.insert("*");
 
-    for (boost::program_options::detail::config_file_iterator it(stream, setOptions), end; it != end; ++it)
-    {
-        std::string strKey = std::string("-") + it->string_key;
-        std::string strValue = it->value[0];
+    auto options = GetConfigOptions(stream);
+
+    for (auto it = options.begin(), end = options.end(); it != end; ++it) {
```
Comment (qmma70) [inline src/util.cpp]: Use `cbegin()` and `cend()` please.

File: src/util.cpp [hunk 3]
```diff
@@ -927,6 +927,37 @@ std::string ArgsManager::GetChainName() const
...
+        if ((pos = str.find('#')) != std::string::npos)
+            str = str.substr(0, pos);
+        str = trimString(str, " \t\r\n");
+        if (!str.empty()) {
+            if (*str.begin() == '[' && *str.rbegin() == ']')
+                prefix = str.substr(1, str.size() - 2) + '.';
+            else if ((pos = str.find('=')) != std::string::npos) {
+                std::string name = prefix + trimString(str.substr(0, pos), " \t\r\n");
+                std::string value = trimString(str.substr(pos + 1), " \t\r\n");
+                options.push_back(std::pair<std::string, std::string>(name, value));
```
Comment (qmma70) [inline src/util.cpp]: Minor: can use `std::make_pair(name, value)` for simplicity.

File: src/util.h (hunk 1 above)
Comment (ken2812221) [reply src/util.h]: No reason. So make it uppercase.

File: src/util.cpp (hunk 2 above)
Comment (ken2812221) [reply src/util.cpp]: I make it range-based

File: src/util.cpp (hunk 3 above)
Comment (ken2812221) [reply src/util.cpp]: `emplace_back` is more simple.

**Comment by qmma70** [comment]
I'm utAck on this. Nicely done.

This could have also been implemented with regex, but the pattern is simple enough that the logic is easily understandable.

File: src/util.cpp [hunk 4]
```diff
@@ -815,13 +814,9 @@ bool ArgsManager::ReadConfigStream(std::istream& stream, std::string& error, boo
...
     LOCK(cs_args);
 
-    std::set<std::string> setOptions;
-    setOptions.insert("*");
-
-    for (boost::program_options::detail::config_file_iterator it(stream, setOptions), end; it != end; ++it)
-    {
-        std::string strKey = std::string("-") + it->string_key;
-        std::string strValue = it->value[0];
+    for (const std::pair<std::string, std::string>& option : GetConfigOptions(stream)) {
```
Comment (Empact) [inline src/util.cpp]: Key type should be `const`

File: src/util.h [hunk 5]
```diff
@@ -282,6 +282,17 @@ class ArgsManager
...
+private:
+    /**
+     * Trim string by the pattern
+     */
+    std::string TrimString(const std::string& str, const std::string& pattern);
+
+    /**
+     * Get config options by reading from stream
+     */
+    std::vector<std::pair<std::string, std::string>> GetConfigOptions(std::istream& stream);
```
Comment (Empact) [inline src/util.h]: Why not a map?

**Comment by Empact** [comment]
Concept ACK - maybe include the tests from #12744?

**Comment by ken2812221** [comment]
@Empact  Those tests have been added by #12713

**Comment by ken2812221** [comment]
Current on master, if there are some options in the config file without `=`, they would cause an exception. In this PR, I simply ignore them, or should I also throw an exception?

**Review by jezzalladmins — CHANGES_REQUESTED** [review]
Run

**Comment by MarcoFalke** [comment]
Update `doc/build-unix.md` as well?

**Comment by ken2812221** [comment]
@MarcoFalke Sure

File: src/util.cpp [hunk 6]
```diff
@@ -927,6 +922,37 @@ std::string ArgsManager::GetChainName() const
...
+
+std::vector<std::pair<std::string, std::string>> ArgsManager::GetConfigOptions(std::istream& stream)
+{
+    std::vector<std::pair<std::string, std::string>> options;
+    std::string str, prefix;
+    std::string::size_type pos;
+    while (std::getline(stream, str)) {
+        if ((pos = str.find('#')) != std::string::npos)
+            str = str.substr(0, pos);
+        str = TrimString(str, " \t\r\n");
```
[the remark this answers is not in the document]
Comment (ken2812221) [reply src/util.cpp]: I'm not sure this is what you want.

File: src/util.cpp [hunk 7]
```diff
@@ -927,6 +922,37 @@ std::string ArgsManager::GetChainName() const
...
+        if (!str.empty()) {
+            if (*str.begin() == '[' && *str.rbegin() == ']')
+                prefix = str.substr(1, str.size() - 2) + '.';
+            else if ((pos = str.find('=')) != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (ken2812221) [reply src/util.cpp]: Fixed

File: src/util.cpp [hunk 8]
```diff
@@ -927,6 +922,37 @@ std::string ArgsManager::GetChainName() const
...
+    return str.substr(front, end - front + 1);
+}
+
+std::vector<std::pair<std::string, std::string>> ArgsManager::GetConfigOptions(std::istream& stream)
+{
+    std::vector<std::pair<std::string, std::string>> options;
+    std::string str, prefix;
+    std::string::size_type pos;
+    while (std::getline(stream, str)) {
+        if ((pos = str.find('#')) != std::string::npos)
```
[the remark this answers is not in the document]
Comment (ken2812221) [reply src/util.cpp]: Fixed

File: src/util.cpp [hunk 9]
```diff
@@ -927,6 +922,40 @@ std::string ArgsManager::GetChainName() const
     return CBaseChainParams::MAIN;
 }
 
+std::string ArgsManager::TrimString(const std::string& str, const std::string& pattern)
```
Comment (sipa) [inline src/util.cpp]: This can just be a static function; it doesn't need to be a method of `ArgsManager`.

**Comment by sipa** [comment]
Why close?

**Comment by ken2812221** [comment]
Sorry, misclicked

**Comment by sipa** [comment]
utACK f447a0a7079619f0d650084df192781cca9fd826

File: build-aux/m4/ax_boost_program_options.m4 [hunk 10]
```diff
@@ -1,108 +0,0 @@
-# ============================================================================
```
Comment (laanwj) [inline build-aux/m4/ax_boost_program_options.m4]: I like this file-bloodbath.

**Comment by laanwj** [comment]
+42 −123 even though this is *removing* an external dependency, if that isn't a clear indication that this should be done I don't know what is.
utACK f447a0a7079619f0d650084df192781cca9fd826

**Pull request merged**